rbd: add `aws-sts-metdata` encryption type

Rakshith-R opened the pull request:
With Amazon STS and kubernetes cluster is configured with
OIDC identity provider, credentials to access Amazon KMS
can be fetched using oidc-token(serviceaccount token).
Each tenant/namespace needs to create a secret with aws region,
role and CMK ARN.
Ceph-CSI will assume the given role with oidc token and access
aws KMS, with given CMK to encrypt/decrypt DEK which will stored
in the image metdata.

Refer: https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html
Resolves: #2879

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: add `aws-sts-metdata` encryption type

With Amazon STS and kubernetes cluster is configured with
OIDC identity provider, credentials to access Amazon KMS
can be fetched using oidc-token(serviceaccount token).
Each tenant/namespace needs to create a secret with aws region,
role and CMK ARN.
Ceph-CSI will assume the given role with oidc token and access
aws KMS, with given CMK to encrypt/decrypt DEK which will stored
in the image metdata.

Refer: https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html
Resolves: #2879

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+// DecryptDEK uses the Amazon KMS and the configured CMK to decrypt the DEK.
+func (kms *awsSTSMetadataKMS) DecryptDEK(volumeID, encryptedDEK string) (string, error) {
```
volumeID is not required for decrypt is that correct?

Madhu-1 commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,238 @@
+	// base64 encode the encrypted DEK, so that storing it should not have
+	// issues
+	encryptedDEK :=
+		base64.StdEncoding.EncodeToString(result.CiphertextBlob)
```
return encodedString directly no need for extra variable

Madhu-1 commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+// EncryptDEK uses the Amazon KMS and the configured CMK to encrypt the DEK.
+func (kms *awsSTSMetadataKMS) EncryptDEK(volumeID, plainDEK string) (string, error) {
```
even for encryption volumeID is not required?

Madhu-1 commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+// getWebIdentityToken returns the web identity token from the file.
+func (kms *awsSTSMetadataKMS) getWebIdentityToken() (string, error) {
```
instead of kms as receiver use `as` km is generic receiver.

Rakshith-R replied:
> instead of kms as receiver use `as` km is generic receiver.


`func (asmkms *awsSTSMetadataKMS) getWebIdentityToken() (string, error) {`

do you mean something like this ?

Madhu-1 replied:
some like below as per https://github.com/golang/go/wiki/CodeReviewComments#receiver-names

```go
func (as *awsSTSMetadataKMS) getWebIdentityToken() (string, error) {
```

Madhu-1 commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,238 @@
+			kms.namespace, kms.secretName, err)
+	}
+
+	config := make(map[string]interface{})
```
use map[string]string? instead of interface as the value is always string?

Madhu-1 commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,238 @@
+	// read the Kubernetes Secret with aws region, role & cmk ARN.
+	secrets, err := kms.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secrets for %T: %w", kms,
```
why %T in the error message? do we need to return type?

Madhu-1 commented on go.mod:
```diff
@@ -40,9 +40,16 @@ require (
 	sigs.k8s.io/controller-runtime v0.11.0-beta.0.0.20211208212546-f236f0345ad2
 )
 
+require github.com/aws/aws-sdk-go-v2/service/sts v1.15.0
```
move this to above require section

Madhu-1 commented on docs/deploy-rbd.md:
```diff
@@ -382,6 +382,31 @@ the AWS KMS is expected to contain:
+   passphrase
+1. `awsRegion`: the region where the AWS STS and KMS service is available.
+
+This Secret is expected to be created by the tenant/user in each namespace where
```
or this secret can be created in the namespace where cephcsi deployed which will be treated as a global one?

Rakshith-R replied:
> or this secret can be created in the namespace where cephcsi deployed which will be treated as a global one?

no, it has to be created in each namespace where pvc will be created.

Madhu-1 commented on docs/deploy-rbd.md:
```diff
@@ -382,6 +382,31 @@ the AWS KMS is expected to contain:
 
+#### Configuring Amazon KMS with Amazon STS
+
+Ceph-CSI can be configured to use [Amazon STS](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html), when kubernetes cluster is configured with OIDC identity provider to fetch
```
reduce the line lenght? split into multiple lines

yati1998 reviewed (commented):
LGTM

Rakshith-R committed:
cleanup: fix csi-kms-connection-details.yaml ibmkeyprotect-test ex

The last option in `ibmkeyprotect-test` example had a extra `,`
at the end, which caused parsing to fail. This commit removes
the same.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented on examples/kms/vault/csi-kms-connection-details.yaml:
```diff
@@ -67,7 +67,12 @@ data:
       "IBM_KP_BASE_URL": "https://us-south.kms.cloud.ibm.com",
       "IBM_KP_TOKEN_URL": "https://iam.cloud.ibm.com/oidc/token",
-      "IBM_KP_REGION": "us-south-2",
+      "IBM_KP_REGION": "us-south-2"
+    }
+  aws-sts-metadata-test: |-
+    {
+      "encryptionKMSType": "aws-sts-metadata",
+      "secretName": "ceph-csi-aws-credentials"
```
Added a small fix here, Please take a look again

humblec commented on docs/deploy-rbd.md:
```diff
@@ -382,6 +382,34 @@ the AWS KMS is expected to contain:
+   AWS. This defaults to `ceph-csi-aws-credentials`.
+
+The [Secret with credentials](../examples/kms/vault/aws-sts-credentials.yaml) for
+the AWS KMS is expected to contain:
```
@Rakshith-R can we mention or give some hint on whats are required/mandatory params ?

Rakshith-R replied:
> @Rakshith-R can we mention or give some hint on whats are required/mandatory params ?

All three below are mandatory required parameters. 
As the sentence suggests so `The Secret with credentials for
the AWS KMS is expected to contain:`

humblec commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,237 @@
+	awsSTSSecretNameKey = "secretName"
+
+	// The following options are part of the Kubernetes Secrets.
+	//
```
do we need this extra line over here @Rakshith-R ?

Rakshith-R replied:
@humblec , This helps with better readability. 
IMO its fine.

humblec commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,237 @@
+	awsSTSRegionKey  = "awsRegion"
+
+	// tokenFilePath is the path to the file containing the OIDC token.
+	//
```
same here?

humblec commented on internal/kms/aws_sts_metadata.go:
```diff
@@ -0,0 +1,237 @@
+	}
+
+	config := make(map[string]string)
+
```
may be good to remove these blank lines.

Rakshith-R replied:
done

humblec reviewed (commented):
ptal.. few minor comments.

humblec commented:
@Rakshith-R just need a rebase, rest looks good to me

Rakshith-R commented:
> @Rakshith-R just need a rebase, rest looks good to me

I'll rebase aftter https://github.com/ceph/ceph-csi/pull/2936 is merged, since that pr may also cause conflicts.
